pkg/gitparse/gitparse.go

```diff
@@ -194,97 +194,123 @@ func (c *Parser) fromReader(ctx context.Context, stdOut io.Reader, commitChan ch
 	var currentDiff *Diff
 
 	defer common.RecoverWithExit(ctx)
+	defer close(commitChan)
 	for {
-		line, err := outReader.ReadBytes([]byte("\n")[0])
-		if err != nil && len(line) == 0 {
-			break
-		}
-		switch {
-		case isCommitLine(line):
-			// If there is a currentDiff, add it to currentCommit.
-			if currentDiff != nil && currentDiff.Content.Len() > 0 {
-				currentCommit.Diffs = append(currentCommit.Diffs, *currentDiff)
-			}
-			// If there is a currentCommit, send it to the channel.
-			if currentCommit != nil {
-				commitChan <- *currentCommit
-			}
-			// Create a new currentDiff and currentCommit
-			currentDiff = &Diff{}
-			currentCommit = &Commit{
-				Message: strings.Builder{},
-			}
-			// Check that the commit line contains a hash and set it.
-			if len(line) >= 47 {
-				currentCommit.Hash = string(line[7:47])
-			}
-		case isAuthorLine(line):
-			currentCommit.Author = strings.TrimRight(string(line[8:]), "\n")
-		case isDateLine(line):
-			date, err := time.Parse(c.dateFormat, strings.TrimSpace(string(line[6:])))
-			if err != nil {
-				log.WithError(err).Debug("Could not parse date from git stream.")
-			}
-			currentCommit.Date = date
-		case isDiffLine(line):
-			// This should never be nil, but check in case the stdin stream is messed up.
-			if currentCommit == nil {
-				currentCommit = &Commit{}
-			}
-			if currentDiff != nil && currentDiff.Content.Len() > 0 {
-				currentCommit.Diffs = append(currentCommit.Diffs, *currentDiff)
-			}
-			currentDiff = &Diff{}
-		case isModeLine(line):
-			// NoOp
-		case isIndexLine(line):
-			// NoOp
-		case isPlusFileLine(line):
-			currentDiff.PathB = strings.TrimRight(string(line[6:]), "\n")
-		case isMinusFileLine(line):
-			// NoOp
-		case isPlusDiffLine(line):
-			currentDiff.Content.Write(line[1:])
-		case isMinusDiffLine(line):
-			// NoOp. We only care about additions.
-		case isMessageLine(line):
-			currentCommit.Message.Write(line[4:])
-		case isContextDiffLine(line):
-			currentDiff.Content.Write([]byte("\n"))
-		case isBinaryLine(line):
-			currentDiff.IsBinary = true
-			currentDiff.PathB = pathFromBinaryLine(line)
-		case isLineNumberDiffLine(line):
-			if currentDiff != nil && currentDiff.Content.Len() > 0 {
-				currentCommit.Diffs = append(currentCommit.Diffs, *currentDiff)
-			}
-			newDiff := &Diff{
-				PathB: currentDiff.PathB,
+		select {
```

Comment:
question: Could we remove this select and replace it w/ `common.IsDone(ctx)` just to reduce the indentation. I think that has the same effect 🤔 
```
if common.IsDone(ctx) {
    cleanupParse(currentCommit, currentDiff, commitChan)
    return
}
```